### ISSUE TITLE
PAPlayer: Gapless play broken for tagged mka files with DTS/Dolby tracks and pass-through

### ISSUE THREAD

**Issue by henkypunky** [xbmc/xbmc#16341]
### Describe the bug
Here is a clear and concise description of what the problem is:

I have a couple of mka files each representing a full surround sound album. The files include original chapter info plus the audio track which can be DTS, DTSMA, AC3, LPCM (actually also TrueHD, however, pass-through is broken in Kodi 18.3). It shows that the files including non-LPCM data are no more played gapless - there is an interruption of about 1 sec between tracks even when the album has smooth transitions between tracks as many concept albums do. On my AVR I can see that the speaker indicators are shortly going off and on again in this moment. I recently upgraded my vero4k from OSMC Oct 2018 baed on Kodi 17 to OSMC June update based on Kodi 18.3. On Kodi 17 all files were played gapless without any issues. mka files containing LPCM audio data are not affected. I have pass-through enabled for all these CODEC types.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

All mka files holding data of any CODEC should play gapless (provided that the album has no pauses between tracks). 

## Actual Behavior

There is a pause of about 1 sec between tracks even when the recording actually has no gaps but smooth transitions between tracks.

## Possible Fix

Sorry, I don't know... However, as it was working in Kodi 17 I guess it's a regression that is most likely easy to fix (referring to the regression in DTSHD pass-through which was also recently fixed in 18.3 and TrueHD pas-through which is about to be fixed in 18.4 with a few lines of code which apparently were falling through the cracks when PAPlayer got an overhaul).

### To Reproduce

1.  Provide a mka file with this characteristics: 
         Holds chapters, one per track
         Holds an audio track encoded with either DTS(MA) or  AC3
2. Ensure pass-through is enabled for said codec
3. Play the file, wait for track transition (you might want to fast-forward to speed up things)

### Debuglog

The debuglog can be found here:
 https://paste.kodi.tv/amewojagis.kodi
Notes:
1) The log includes tests with 4 mka files holding different audio tracks (DTSMA, AC3, DTS96/24, LPCM)
2) To safe time I did some fast forward to reach the end of a track faster. For album 2 I 
3) I search&replaced the original paths and filenames by "/music/album[1-4]<codectype>.mka" for privacy reasons 
4) I have cut a couple of thousand lines in the beginning (see marker) - there was a scheduled backup running right before I started my testing
accidentally went too far so I had to go back again..

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [X ] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: OSMC June 2019 on vero 4k
 - Kodi version: 18.3

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by fritsch** [comment]
Gappless passthrough is not supported. Sink has to be exclusively opened. Disable passthrough if you want gappless and set audio to optimized.

**Comment by henkypunky** [comment]
Well, when I playback the same album as mkv (i.e. the video track with still images still included) via the video player, it also manages to play it gapless - how this? And again, I would bet on it was working with Kodi 17 - maybe with the help of an OSMC / vero4k specific patch?. But I have to say that I had no tags included in the mka but a cue sheet file at that time. With Kodi 18 the cue file was no more respected so had to incorporate tags into the mka file. Maybe related to this? I will revert back to Kodi 17 quite soon just to ensure I am not talking nonsense.

**Comment by FernetMenta** [comment]
It seemed to work by accident in v17 because the audio sink was not reconfigured when codec of passthrough changed.

**Comment by henkypunky** [comment]
Cool, so this proves it is possible in principle ;-) Why not re-applying this "accident" in Kodi 18? ;-)

**Comment by fritsch** [comment]
Easy answer: if PT format changes (e.g. DTS to AC3) but sink format, e.g. 16 bit digital iec, the next track won't play at all. Even if you press "next" on e.g. DTS, while playing AC3 -> you get a problem.

**Comment by fritsch** [comment]
See: a28e0c5e5250b5b683f7672f172107620ee13745

**Comment by henkypunky** [comment]
I see, however, we are talking about playing a single mka file which is just logically partitioned by chapters - it is a single, continous audio track for the length of the album so how should codec change in between?

**Comment by fritsch** [comment]
No, it should not - also above code does not. Make sure your audio is not configured to "Best Match" - as what you describe exactly sounds like it.

**Comment by henkypunky** [comment]
Indeed, wascset to Best Match. Switched to Optimized but unfortunately same situation. However, I noticed that some track transitions are gapless while most of them are not (I mean within the same album). I now have also noticed that lpcm codec is not perfect as there are slight hickups on track change.

**Comment by fritsch** [comment]
Please add all debuglogs showing your issue. So that a future dev that might pick it up has all information he needs.

Also produce the logs with a platform that does not patch key components, please. Aml/Vero uses a patched Alsa sink with kernel hacks to achieve multi channel and stuff. All of them can interfere with that behaviour.

**Comment by henkypunky** [comment]
Ok, will do. I have a Raspi 3 available - is this ok? However, it does not support pass-through of HD formats. But as the issue also shows with normal DTS/AC3 I think it's fine. When you say "all debuglogs", what do you mean? More detailed logs as I already attached? If yes, can you give me a hint how to create these more detailed logs? Thanks!

**Comment by fritsch** [comment]
That was meant like: if you test new things, please add a debuglog for those, especially if you change audio settings.

Thanks in advance.

So what should work: if track a is dts and track 2 is dts we expect it continues gappless. Btw. it also needs to stay same DTS, e.g. 1024 vs 2048 etc.

**Comment by henkypunky** [comment]
I see. So in the meantime I also tested with Kodi 18.2 on FireTV Stick 4k - same results. However, I have noticed that a few trackchanges are gapless - repeatable; always the same track transitions. I was suspecting the chapter info probably has minimum gaps between end time of track n and start time of track n+1 but this is not the case at all. Start time of next track always matches exactly end time of previous one. 

Just to be 100% clear what the issue is about: 
- It is not about playing multiple mka files in sequence
- It is about mka files representing a full album of music with smooth transitions of successive tracks
- The mka files result from mka rips with MakeMKV with the video track just removed afterwards using Mktoolnix
- The result is a mka file containing a) the original audio track as a single consecutive stream which is naturally encoded in a single format, b) original chapter info which corresponds to the track titles, c) tags for artist, album name, track titles
- The video player plays the original mkv (still including the video stream) flawlessly - no gaps
- The audio player shows the gaps even though the audio stream and the chapter info is exactly the same in both mka and mkv file

Not sure, do you have any mkv at hand (I guess it could be also a movie) which you could turn into a mka by removing the video stream? If so, you should be able to reproduce by your own.

**Comment by fritsch** [comment]
I understood you well. I don't plan to fix it though, just want to make sure that IF someone fixes it all logs are properly documented.

**Comment by jjd-uk** [comment]
As a workaround could be to create a **playercorefactory.xml** file to make mka files use videoplayer instead of paplayer.

I believe the code xml would go something like:

` <playercorefactory>
  <players>
<player name="VideoPlayer" audio="true" video="false" />
    </player>
  </players>
  <rules action="prepend">
    <rule filetypes="mka" player="VideoPlayer"/>
  </rules>
 </playercorefactory>`

Normally **playercorefactory.xml** is used for adding external player so it's documented on the wiki at https://kodi.wiki/view/External_players

**Comment by henkypunky** [comment]
@jjd-uk Thanks for that, will try.

**Comment by henkypunky** [comment]
I now was able to reproduce the issue on a Raspberry PI3 running Libreelec 9.0.2 with Kodi 18.2:

https://paste.kodi.tv/rezetijuke.kodi

The log document playing the first tracks of a album represented by a single mka file containing continuous AC3 audio stream. To speed up things I was seeking forward couple of times to get close to the end of tracks.
This proves,to me that the issue is not OSMC/Vero4k related. 

I hope this helps to fix the issue. If further info / logs / testing is needed just let me know! I am dying for a fix as I have to downgrade to Kodi 17 otherwise. Thanks!

**Comment by fritsch** [comment]
Looking at your log, I see exactly ones:

```
2019-07-06 09:37:20.783 T:1906566000   DEBUG: CActiveAESink::OpenSink - SinkPi Initialized:
2019-07-06 09:37:20.783 T:1906566000   DEBUG:   Output Device : HDMI
2019-07-06 09:37:20.784 T:1906566000   DEBUG:   Sample Rate   : 48000
2019-07-06 09:37:20.784 T:1906566000   DEBUG:   Sample Format : AE_FMT_S16NE
2019-07-06 09:37:20.784 T:1906566000   DEBUG:   Channel Count : 2
2019-07-06 09:37:20.784 T:1906566000   DEBUG:   Channel Layout: RAW, RAW
2019-07-06 09:37:20.784 T:1906566000   DEBUG:   Frames        : 2400
2019-07-06 09:37:20.784 T:1906566000   DEBUG:   Frame Size    : 4
2019-07-06 09:37:20.797 T:1914958704   DEBUG: CActiveAE::ClearDiscardedBuffers - buffer pool deleted
```

Means, the Sink is not destroyed through all your testing. I would have expected that we see "reconfiguration" of the sink, but that's not it.

**Comment by henkypunky** [comment]
But it says
`2019-07-06 09:37:20.581 T:1937355424   DEBUG: Loading additional tag info for file smb://synds//music/album_ac3.mka
2019-07-06 09:37:20.696 T:1583022960   DEBUG: CSMBFile::Open - opened smb://USERNAME:PASSWORD@synds//music/album_ac3.mka, fd=10000
2019-07-06 09:37:20.702 T:1937355424   DEBUG: CSMBFile::Open - opened smb://USERNAME:PASSWORD@synds//music/album_ac3.mka, fd=10001`
for each new track. Maybe the extra open of the file interferes with the playback? Just guessing.

**Comment by FernetMenta** [comment]
I suggest that you record a log without doing seeks. Those seeks just pollute the log with distracting info.

**Comment by henkypunky** [comment]
Ok, this is same album again playing tracks 1, 2, beginning of 3 without any seeks this time:

https://paste.kodi.tv/avaqequmut.kodi

I then did another experiment as I suspect the embedded tags. I removed the tags again with MKVToolNix so that the mka only holds chapter info plus teh actual audio stream.  This is the log playing tracks 1, 2, 3, beginnign of 4:

https://paste.kodi.tv/oxadufomow.kodi

Well, I have to say it plays w/o gaps! However, I have got the feeling that there is kind of hickup between trackls, i.e. some 100 ms skipped... However, not 100% sure yet. But for sure, it makes a difference. And when I was on Kodi 17, my files had no embedded tags but an extra cue sheet file. I added the tags recently as I was suspecting that Kodi 18 no more respects the cue sheet file (is it true by the way?).

**Comment by FernetMenta** [comment]
What makes you think cue file support was dropped? PAPlayer does not reopen a music file if the next track to play is the next track of the same file. That ensures gapless. This is not true for tags.

I don't know much about tags. Always found it strange how it reads the information of an already opened file. Won't surprise me if that interferes with playback.

**Comment by henkypunky** [comment]
Well, the cue file is just ignored on scanning. It seems to extract solely info from mka and if it lacks any info about album or artist, the album will not show up in your album at all. The .cue files have the same name as the according .mka file and are colocated in the same folder. This was working finw with Kodi 17 but is no more. From the logs I can see that the .cue file is considered, though. I would be happy if the cue file would be considered again in Kodi 18 as I would just wipe the tags from my mka files again and that's it. So maybe I should rather open a new defect related to the cue file ignorance?
However, I am wondering why the tags are read from the file over and over again during playback at all... Shouldn't it e read only once at scan time and by that is transferred to the database?

**Comment by fritsch** [comment]
cue files are working perfectly, just tested it. Scanned one to library, all tags appear and I can play one after the other.

I used an mp3 file with corresponding cue file - that was provided by a user some time ago.

**Comment by DaveTBlake** [comment]
As discussed on forum and issue https://github.com/xbmc/xbmc/issues/16349 Kodi has a problem with scanning some tags from .mka files because ffmpeg has a bug in Matroska tag reading. In particular album tag is not read by ffmpeg hence no album entry is made in the music library, and so attempts to scrape an nfo file for that album will not work either - nfo files only add additional information to the albums in the music database, not make new ones.

However I don't think this should impact gapless playback, just what items get added to the library.

**Comment by henkypunky** [comment]
> cue files are working perfectly, just tested it. Scanned one to library, all tags appear and I can play one after the other.
> 
> I used an mp3 file with corresponding cue file - that was provided by a user some time ago.
> 

Well, yes, for mp3 it's working but not for mka anymore. I just opened a new issue on this: 
https://github.com/xbmc/xbmc/issues/16359

**Comment by henkypunky** [comment]
Sorry, closed this issue by accident... It is still valid, however there are 2 aspects:
1) PAPplayer shows gaps between tracks as soon as mka file contains tags (this issue)
2) cue sheet files for mka files are ignored (new issue: https://github.com/xbmc/xbmc/issues/16359)

**Comment by fritsch** [comment]
You are wrong. I created a cue file, see below:
```
fritsch@jabba:/mnt/data/tosort/demo_files/cue2$ cat file.cue 
PERFORMER "Various Artists"
TITLE "ASDFWatergate 11 - Solomun"
FILE "file.mka" MKA
  TRACK 01 AUDIO
    TITLE "ASDFRespect"
    PERFORMER "Alliance Ethnik"
    INDEX 01 00:00:00
  TRACK 02 AUDIO
    TITLE "ASDFDon't Mess With My Man"
    PERFORMER "Lucy Pearl"
    INDEX 01 04:02:43
  TRACK 03 AUDIO
    TITLE "ASDFKackvogel"
    PERFORMER "Solomun"
    INDEX 01 06:59:10
  TRACK 04 AUDIO
```

I used the old mp3 I had and ffmpeg -i file.mp3 file.mka - It succesfully scanned into my library and it also plays.

So 2. is not a problem I can reproduce.

**Comment by fritsch** [comment]
See:
```
2019-07-07 18:29:51.762 T:140624353457920   DEBUG: CNFSFile::Open - opened mnt/data/tosort/demo_files/cue2/file.mka
2019-07-07 18:29:51.847 T:140624353457920   DEBUG: CNFSFile::Close closing file mnt/data/tosort/demo_files/cue2/file.mka
2019-07-07 18:29:51.850 T:140624353457920   DEBUG: DoScan Scanning dir 'nfs://192.168.178.29/mnt/data/tosort/demo_files/cue2/' as not in the database
2019-07-07 18:29:51.854 T:140624353457920   DEBUG: CNFSFile::Open - opened mnt/data/tosort/demo_files/cue2/file.cue
2019-07-07 18:29:51.865 T:140624353457920   DEBUG: CNFSFile::Close closing file mnt/data/tosort/demo_files/cue2/file.cue
2019-07-07 18:29:51.884 T:140624353457920   DEBUG: CNFSFile::Open - opened mnt/data/tosort/demo_files/cue2/file.mka
2019-07-07 18:29:51.940 T:140624353457920   DEBUG: CNFSFile::Close closing file mnt/data/tosort/demo_files/cue2/file.mka
2019-07-07 18:29:51.945 T:140624353457920   DEBUG: Album 'ASDFWatergate 11 - Solomun' is a compilation as there's no overlapping tracks and the album artist is 'Various'
```

Please close the bug - as we currently track here something else than what we originally started and confusion level is high.

**Comment by DaveTBlake** [comment]
> So maybe I should rather open a new defect related to the cue file ignorance?

No @henkypunky that is premature. I suggest that you start a thread on the music support sub-forum, providing example .mka file and cuesheet that shows the issue

**Comment by henkypunky** [comment]
@fritsch Thanks for the info. I was able to find the difference(s) between your mka/cue file and mine:
1) You are using "MKA" as the type in the FILE statement (however, it did not make any difference when I set it also to MKA in my cue file)
2) I suppose your mka did not include chapter info but my files do. I deleted it from the mka with mkvtoolnix and voila - the cue file is respected and the album shows in the library!
@DaveTBlake 
However, my files were working fine with Kodi 17 - I did not touch the files when upgrading to Kodi 18.3 on my vero4k. So I am wondering: Was I just lucky that it was working with Kodi 17 or is it a regression that the mka files must no more include chapter info?

**Comment by rvbgnu** [comment]
Sorry for reviving this 2 years old issue. I have the gapless problem on simple MP3 audio files, typically dj sets or live concerts. Each track is an MP3 track, same from the original CD, or official download. Coming from iTunes on Mac, I have years of different files format and versions, and I could tested a lot of them.

My Kodi install is fresh on Mac, v18.5. Settings have been restore to default for these tests. Library on local hard drive, same as iTunes.

I found out that every track encoded with LAME 3.98 or more, with id3 tags v2.3, plays very well, and gapless. Same for bitrate 160 kbps, 192 or 320. But tracks encoded by iTunes, v3 to v12, with tags v 2.2 or v2.3, don't play correctly, with breaks in between the next one.

I hope it helps. I may post this on a more relevant location, if needed.

**Comment by VonMagnum** [comment]
Since v21 Omega of Kodi, it shows the last track first on MKA files here on my Nvidia Shield, which accidentally auto upgraded and you cannot play that track directly, only indirectly from the next to last track. I've been unable to get a single person to verify if this happens to them too on the KODI forums despite 3 attempts and seeing a Team KODI member reply to someone else in the same thread. No one even replies. I use videoplayer to avoid gaps in playback (only issue is the time doesn't reset on the next track but keeps incrementing), but changing it back to paplayer made no difference. It worked fine in v19 and v20 versions of KODI.

**Comment by the-black-eagle** [comment]
Please post a link to the relevant forum thread as that's most likely the best way to get this looked at. Cheers.

**Comment by VonMagnum** [comment]
These are the four attempts to find out if anyone else had the same MKA listing issue in Omega.  No replies to any of them.  Fritsch posted right after the first one to someone else.  I don't know if someone wanted a log but doesn't want to say so.  I was just trying to verify if anyone else even saw the problem.  I got the feeling NO ONE uses .MKA files.  I keep the full MKV files (they work fine in Omega, but usually have bright white backgrounds and take up more space on the Shield hard drive so I use MKA files on the Shield's own drive for fastest playback regardless of whether my server is doing anything else (I keep the MKVs on the server).  MKA files give me a dark room and take up considerably less space).

https://forum.kodi.tv/showthread.php?tid=377253&pid=3196336#pid3196336

https://forum.kodi.tv/showthread.php?tid=376494&pid=3196506#pid3196506

https://forum.kodi.tv/showthread.php?tid=377275&pid=3196752#pid3196752

https://forum.kodi.tv/showthread.php?tid=372668&pid=3205629#pid3205629

**Comment by the-black-eagle** [comment]
OK, well jumping into random threads about other things isn't the way to get support.  I have split off one of your posts to create a new thread where we can discuss it properly https://forum.kodi.tv/showthread.php?tid=378468  As such I'll close this for now until we determine if it is actually a bug or not.

Thanks.
